pandas/core/indexes/base.py

```diff
@@ -5534,7 +5544,12 @@ def _should_fallback_to_positional(self) -> bool:
         """
         Should an integer key be treated as positional?
         """
-        return not self.holds_integer()
+        if is_integer_dtype(self):
+            return False
+        elif is_object_dtype(self):
+            return self.inferred_type not in ["integer", "mixed-integer"]
+        else:
+            return True
```

Comment:
Do we want to just use `return self.inferred_type not in ["integer", "mixed-integer"]` here?

Reply:
looks like all our tests pass if we change this whole thing to `return self.dtype.kind in ["i", "u"]`.  I think the thing to do is drop this for now and deprecate it in 2.x and change it to this simpler check in 3.0.

Reply:
I'm not sure I follow how could deprecate this, as this is an internal attribute, and I don't think this changes behavior (though I'm not 100 % sure there isn't an edge case).

Are you saying only use `return self.inferred_type not in ["integer", "mixed-integer"]` here and drop the if/elif/else?

Reply:
> though I'm not 100 % sure there isn't an edge case

Under my proposal, `pd.Index([1, 2], dtype=object).holds_integer()` would change in the future.

The only places where we use holds_integer are in _should_fallback_to_positional (which I expect to get rid of in 3.0) and plotting code.  It would require some actual effort to determine if the suggested change could affect users via the plotting code.

Reply:
Concretely in this PR, I git difficulty understanding what you want. Can you say what code you want changed, e.g. should the proposed deprecation of `holds_integer` be dropped? IMO it does seems like a good idea to deprecate the method...

Reply:
in this PR i would suggest

```
def _holds_integer(self) -> bool:
     return self.inferred_type not in ["integer", "mixed-integer"] 

def holds_integer(self) -> bool:
    warngs.warn("holds_integer is deprecated [...]", FutureWarning)
    return self._holds_integer()
```

Reply:
Ok, nop problem, I've just updated the PR.